src/conditions/datadog_search.rs

```diff
@@ -27,215 +26,297 @@ impl_generate_config_from_default!(DatadogSearchConfig);
     fn check(&self, e: &Event) -> bool {
-        self.matcher.run(e)
+        if let Event::Log(log) = e {
+            EventFilter::run(&self.matcher, log)
+        } else {
+            false
+        }
     }
 }
 
 #[typetag::serde(name = "datadog_search")]
 impl ConditionConfig for DatadogSearchConfig {
     fn build(&self, _enrichment_tables: &enrichment::TableRegistry) -> crate::Result<Condition> {
         let node = parse(&self.source)?;
-        let matcher = as_log(build_matcher(&node, &EventFilter::default()));
+        let matcher = fast_matcher::build_matcher(&node, &EventFilter::default());
 
         Ok(Condition::DatadogSearch(DatadogSearchRunner { matcher }))
     }
 }
 
-/// Run the provided `Matcher` when we're dealing with `LogEvent`s. Otherwise, return false.
-fn as_log(matcher: Box<dyn Matcher<LogEvent>>) -> Box<dyn Matcher<Event>> {
-    Run::boxed(move |ev| match ev {
-        Event::Log(log) => matcher.run(log),
-        _ => false,
-    })
-}
-
 //------------------------------------------------------------------------------
 
 #[derive(Default, Clone)]
-struct EventFilter;
+pub struct EventFilter;
 
-/// Uses the default `Resolver`, to build a `Vec<Field>`.
-impl Resolver for EventFilter {}
-
-impl Filter<LogEvent> for EventFilter {
-    fn exists(&self, field: Field) -> Box<dyn Matcher<LogEvent>> {
-        match field {
-            Field::Tag(tag) => {
-                let starts_with = format!("{}:", tag);
-
-                any_string_match("tags", move |value| {
-                    value == tag || value.starts_with(&starts_with)
-                })
-            }
-            // Literal field 'tags' needs to be compared by key.
-            Field::Reserved(field) if field == "tags" => {
-                any_string_match("tags", move |value| value == field)
-            }
-            Field::Default(f) | Field::Facet(f) | Field::Reserved(f) => {
-                Run::boxed(move |log: &LogEvent| log.get(&f).is_some())
-            }
+impl EventFilter {
+    pub fn run(matcher: &fast_matcher::FastMatcher, log: &LogEvent) -> bool {
+        match &matcher.mode {
+            Mode::One(op) => exec(op, log),
+            Mode::Any(ops) => ops.iter().any(|op| exec(op, log)),
+            Mode::All(ops) => ops.iter().all(|op| exec(op, log)),
         }
     }
+}
 
-    fn equals(&self, field: Field, to_match: &str) -> Box<dyn Matcher<LogEvent>> {
-        match field {
-            // Default fields are compared by word boundary.
-            Field::Default(field) => {
-                let re = word_regex(to_match);
-
-                string_match(&field, move |value| re.is_match(&value))
-            }
-            // A literal "tags" field should match by key.
-            Field::Reserved(field) if field == "tags" => {
-                let to_match = to_match.to_owned();
+fn exec(op: &Op, log: &LogEvent) -> bool {
+    match op {
+        Op::True => true,
+        Op::False => true,
```

Comment:
Uh, really?

Reply:
I'm honestly not totally sure how to interpret this -- my work here has been more or less mechanical translation -- but if this if flipped the other way the languages benchmark tests start to fail, so there is some justification for this. I've gone back and checked the original code here and the `QueryNode::MatchNoDocs`, `MatchAllDocs` works this way too, I think. Perhaps a bug we've preserved? There _is_ very little test coverage for these fields.

Reply:
Hmm, it is surprising to me that both the `True` and `False` operations should return the value `true`, and so it looked very much like a copy-and-paste error. I would look very suspiciously at test cases that fail from the second case being changed to `false` (here and the one I identified below), but I guess it is conceivable this is the way it's supposed to work.